### ISSUE TITLE
Uncaught Exception: TypeError: Cannot read properties of undefined (reading 'listeners')

### ISSUE THREAD

**Issue by zslp2** [cronicle-edge/cronicle-edge#61]
I am getting this error when starting Cronicle on Windows. No errors running bundle.ps1 or configuring storage. Any help will be appreciated.
```powershell
(cronicle) PS C:\ProgramData\CronicleEdge> node .\bin\cronicle.js --echo --foreground --manager --color
[1696946827.091][2023-10-10 10:07:07][servername][8440][Cronicle][debug][2][Cronicle v1.7.1 Starting Up][{"pid":8440,"ppid":9168,"node":"v18.17.1","arch":"x64","platform":"win32","argv":["C:\\ProgramData\\miniforge3\\envs\\cronicle\\node.exe","C:\\ProgramData\\CronicleEdge\\bin\\cronicle.js","--echo","--foreground","--manager","--color"],"execArgv":[]}]
[1696946827.103][2023-10-10 10:07:07][servername][8440][Cronicle][debug][2][Server IP: x.x.x.x, Daemon PID: 8440][]
[1696946827.105][2023-10-10 10:07:07][servername][8440][Cronicle][debug][3][Starting component: Storage][]
[1696946827.107][2023-10-10 10:07:07][servername][8440][Storage][debug][5][Setting up storage system v3.1.16][]
[1696946827.126][2023-10-10 10:07:07][servername][8440][Filesystem][debug][2][Setting up filesystem storage][]
[1696946827.129][2023-10-10 10:07:07][servername][8440][Filesystem][debug][3][Base directory: data][]
[1696946827.13][2023-10-10 10:07:07][servername][8440][Cronicle][debug][3][Starting component: WebServer][]
[1696946827.131][2023-10-10 10:07:07][servername][8440][WebServer][debug][2][pixl-server-web v1.7.1 starting up][]
[1696946827.142][2023-10-10 10:07:07][servername][8440][WebServer][debug][2][Starting HTTP server on port: 3012][]
[1696946827.146][2023-10-10 10:07:07][servername][8440][WebServer][debug][3][Now listening for HTTP connections][{"address":"::","family":"IPv6","port":3012}]
[1696946827.147][2023-10-10 10:07:07][servername][8440][Cronicle][debug][3][Starting component: API][]
[1696946827.153][2023-10-10 10:07:07][servername][8440][API][debug][3][API service listening for base URI: /api][]
[1696946827.154][2023-10-10 10:07:07][servername][8440][WebServer][debug][3][Adding custom URI handler: /\/api\/(\w+)/: API][]
[1696946827.156][2023-10-10 10:07:07][servername][8440][Cronicle][debug][3][Starting component: User][]
[1696946827.156][2023-10-10 10:07:07][servername][8440][User][debug][3][User Manager starting up][]
[1696946827.158][2023-10-10 10:07:07][servername][8440][API][debug][3][Adding API namespace: user][]
[1696946827.158][2023-10-10 10:07:07][servername][8440][Cronicle][debug][3][Starting component: Cronicle][]
[1696946827.16][2023-10-10 10:07:07][servername][8440][Cronicle][debug][3][Cronicle engine starting up][]
[1696946827.211][2023-10-10 10:07:07][servername][8440][API][debug][3][Adding API namespace: app][]
[1696946827.212][2023-10-10 10:07:07][servername][8440][WebServer][debug][3][Adding custom URI filter: /^\/api\/app\/\w+/: API Filter][]
[1696946827.215][2023-10-10 10:07:07][servername][8440][WebServer][debug][3][Adding custom URI handler: /^\/db$/i: Reports][]
[1696946827.216][2023-10-10 10:07:07][servername][8440][WebServer][debug][3][Adding custom URI handler: /^\/console$/i: Console][]
[1696946827.217][2023-10-10 10:07:07][servername][8440][WebServer][debug][3][Adding custom request method handler: OPTIONS: CORS Preflight][]
[1696946827.229][2023-10-10 10:07:07][servername][8440][Cronicle][debug][1][Uncaught Exception: TypeError: Cannot read properties of undefined (reading 'listeners')][TypeError: Cannot read properties of undefined (reading 'listeners')     at _Server.attach (C:\ProgramData\CronicleEdge\bin\cronicle.js:88154:35)     at attach (C:\ProgramData\CronicleEdge\bin\cronicle.js:89281:14)     at _Server.initEngine (C:\ProgramData\CronicleEdge\bin\cronicle.js:92994:43)     at _Server.attach (C:\ProgramData\CronicleEdge\bin\cronicle.js:92945:14)     at __construct.startSocketListener (C:\ProgramData\CronicleEdge\bin\cronicle.js:97209:17)     at __construct.startup (C:\ProgramData\CronicleEdge\bin\cronicle.js:105838:14)     at C:\ProgramData\CronicleEdge\bin\cronicle.js:10306:18     at C:\ProgramData\CronicleEdge\bin\cronicle.js:1328:18     at replenish (C:\ProgramData\CronicleEdge\bin\cronicle.js:438:15)     at iterateeCallback (C:\ProgramData\CronicleEdge\bin\cronicle.js:422:15)]
TypeError: Cannot read properties of undefined (reading 'listeners')
    at _Server.attach (C:\ProgramData\CronicleEdge\bin\cronicle.js:88154:35)
    at attach (C:\ProgramData\CronicleEdge\bin\cronicle.js:89281:14)
    at _Server.initEngine (C:\ProgramData\CronicleEdge\bin\cronicle.js:92994:43)
    at _Server.attach (C:\ProgramData\CronicleEdge\bin\cronicle.js:92945:14)
    at __construct.startSocketListener (C:\ProgramData\CronicleEdge\bin\cronicle.js:97209:17)
    at __construct.startup (C:\ProgramData\CronicleEdge\bin\cronicle.js:105838:14)
    at C:\ProgramData\CronicleEdge\bin\cronicle.js:10306:18
    at C:\ProgramData\CronicleEdge\bin\cronicle.js:1328:18
    at replenish (C:\ProgramData\CronicleEdge\bin\cronicle.js:438:15)
    at iterateeCallback (C:\ProgramData\CronicleEdge\bin\cronicle.js:422:15)
```

Some additional debug information.
```powershell
(cronicle) PS C:\temp\cronicle-edge-1.7.2> node --version
v18.17.1
(cronicle) PS C:\temp\cronicle-edge-1.7.2> npm --version
9.6.7
(cronicle) PS C:\temp\cronicle-edge-1.7.2> npm ls
cronicle-edge@1.7.1 C:\temp\cronicle-edge-1.7.2
+-- @iarna/toml@2.2.5
+-- @mdi/font@2.1.99
+-- activedirectory2@2.2.0
+-- ansi_up@5.2.1
+-- async@2.6.4
+-- bcrypt-node@0.1.0
+-- chart.js@2.9.4
+-- codemirror@5.65.15
+-- diff@5.1.0
+-- dockerode@3.3.5
+-- dotenv@8.6.0
+-- font-awesome@4.7.0
+-- glob@5.0.15
+-- graphlib@2.1.8
+-- jquery-datetimepicker@2.5.21
+-- jquery-ui-dist@1.13.2
+-- jquery@3.7.1
+-- js-yaml@4.1.0
+-- jsonlint-mod@1.7.6
+-- jstimezonedetect@1.0.6
+-- moment-timezone@0.5.37
+-- moment@2.29.4
+-- netmask@2.0.2
+-- openssl-wrapper@0.3.4
+-- pixl-args@1.0.4
+-- pixl-boot@2.0.0
+-- pixl-class@1.0.3
+-- pixl-cli@1.0.16
+-- pixl-config@1.0.13
+-- pixl-json-stream@1.0.8
+-- pixl-logger@2.0.2
+-- pixl-mail@1.0.13
+-- pixl-perf@1.0.9
+-- pixl-request@1.0.37
+-- pixl-server-api@1.0.3
+-- pixl-server-storage@3.1.16
+-- pixl-server-web@1.3.22
+-- pixl-server@1.0.41
+-- pixl-tools@1.0.35
+-- pixl-unit@1.0.14
+-- pixl-webapp@2.0.2
+-- read-last-lines@1.8.0
+-- shell-quote@1.7.3
+-- socket.io-client@4.7.1
+-- socket.io@4.7.1
+-- ssh2@1.14.0
+-- systeminformation@5.21.11
+-- uglify-js@3.14.3
+-- uncatch@1.0.3
+-- vis-network@9.1.7
`-- xss@1.0.14
```

**Comment by mikeTWC1984** [comment]
Looks like crash occurs when UDP listener trying to start up. Is port 3014 is used by something on your machine?

Try to rebuild cronicle again with -Dev parameter (so it won't minimize code in cronicle.js), then paste here +/- 20 lines around the line where exception was thrown ( get it from the stack trace, e.g. form above:
at _Server.attach (C:\ProgramData\CronicleEdge\bin\cronicle.js:88154:35)
)

**Comment by mikeTWC1984** [comment]
One more exercise you can do - in lib/engine.js file - locate this line:
this.setupDiscovery();

and try to comment it out and rebuild cronicle, see if it helps

**Comment by zslp2** [comment]
Ports listening for connections; no conflicts for 3014 or 3012.
```powershell
(cronicle) C:\ProgramData\CronicleEdge>netstat -n -a |findstr LISTENING
  TCP    0.0.0.0:135            0.0.0.0:0              LISTENING
  TCP    0.0.0.0:445            0.0.0.0:0              LISTENING
  TCP    0.0.0.0:2701           0.0.0.0:0              LISTENING
  TCP    0.0.0.0:3389           0.0.0.0:0              LISTENING
  TCP    0.0.0.0:5985           0.0.0.0:0              LISTENING
  TCP    0.0.0.0:8005           0.0.0.0:0              LISTENING
  TCP    0.0.0.0:28002          0.0.0.0:0              LISTENING
  TCP    0.0.0.0:47001          0.0.0.0:0              LISTENING
  TCP    0.0.0.0:49664          0.0.0.0:0              LISTENING
  TCP    0.0.0.0:49665          0.0.0.0:0              LISTENING
  TCP    0.0.0.0:49666          0.0.0.0:0              LISTENING
  TCP    0.0.0.0:50758          0.0.0.0:0              LISTENING
  TCP    0.0.0.0:60943          0.0.0.0:0              LISTENING
  TCP    0.0.0.0:60944          0.0.0.0:0              LISTENING
  TCP    0.0.0.0:61900          0.0.0.0:0              LISTENING
  TCP    0.0.0.0:61903          0.0.0.0:0              LISTENING
  TCP    0.0.0.0:63456          0.0.0.0:0              LISTENING
  TCP    0.0.0.0:63488          0.0.0.0:0              LISTENING
  TCP    127.0.0.1:64589        0.0.0.0:0              LISTENING
  TCP    [::]:135               [::]:0                 LISTENING
  TCP    [::]:445               [::]:0                 LISTENING
  TCP    [::]:2701              [::]:0                 LISTENING
  TCP    [::]:3389              [::]:0                 LISTENING
  TCP    [::]:5985              [::]:0                 LISTENING
  TCP    [::]:8005              [::]:0                 LISTENING
  TCP    [::]:28002             [::]:0                 LISTENING
  TCP    [::]:47001             [::]:0                 LISTENING
  TCP    [::]:49664             [::]:0                 LISTENING
  TCP    [::]:49665             [::]:0                 LISTENING
  TCP    [::]:49666             [::]:0                 LISTENING
  TCP    [::]:50758             [::]:0                 LISTENING
  TCP    [::]:60943             [::]:0                 LISTENING
  TCP    [::]:60944             [::]:0                 LISTENING
  TCP    [::]:61900             [::]:0                 LISTENING
  TCP    [::]:61903             [::]:0                 LISTENING
  TCP    [::]:63456             [::]:0                 LISTENING
  TCP    [::]:63488             [::]:0                 LISTENING
``` 

Pasting bin\cronicle.js:88154 +/- 20 lines.
```javascript
        }
        function onUpgradeError() {
          debug("websocket error before upgrade");
        }
        __name(onUpgradeError, "onUpgradeError");
      }
      /**
       * Captures upgrade requests for a http.Server.
       *
       * @param {http.Server} server
       * @param {Object} options
       * @api public
       */
      attach(server2, options = {}) {
        const path = this._computePath(options);
        const destroyUpgradeTimeout = options.destroyUpgradeTimeout || 1e3;
        function check(req) {
          return path === req.url.slice(0, path.length);
        }
        __name(check, "check");
        const listeners = server2.listeners("request").slice(0); //line 88154
        server2.removeAllListeners("request");
        server2.on("close", this.close.bind(this));
        server2.on("listening", this.init.bind(this));
        server2.on("request", (req, res) => {
          if (check(req)) {
            debug('intercepting request for path "%s"', path);
            this.handleRequest(req, res);
          } else {
            let i2 = 0;
            const l = listeners.length;
            for (; i2 < l; i2++) {
              listeners[i2].call(server2, req, res);
            }
          }
        });
        if (~this.opts.transports.indexOf("websocket")) {
          server2.on("upgrade", (req, socket, head) => {
            if (check(req)) {
              this.handleUpgrade(req, socket, head);
            } else if (false !== options.destroyUpgrade) {
``` 

Exactly the same behavior/error after commenting out lib\engine.js:157 and rebuilding with `.\bundle.ps1 C:\ProgramData\CronicleEdge -Tools -Dev`.
```powershell
TypeError: Cannot read properties of undefined (reading 'listeners')
    at _Server.attach (C:\ProgramData\CronicleEdge\bin\cronicle.js:88154:35)
    at attach (C:\ProgramData\CronicleEdge\bin\cronicle.js:89281:14)
    at _Server.initEngine (C:\ProgramData\CronicleEdge\bin\cronicle.js:92994:43)
    at _Server.attach (C:\ProgramData\CronicleEdge\bin\cronicle.js:92945:14)
    at __construct.startSocketListener (C:\ProgramData\CronicleEdge\bin\cronicle.js:97209:17)
    at __construct.startup (C:\ProgramData\CronicleEdge\bin\cronicle.js:105838:14)
    at C:\ProgramData\CronicleEdge\bin\cronicle.js:10306:18
    at C:\ProgramData\CronicleEdge\bin\cronicle.js:1328:18
    at replenish (C:\ProgramData\CronicleEdge\bin\cronicle.js:438:15)
    at iterateeCallback (C:\ProgramData\CronicleEdge\bin\cronicle.js:422:15)
```

**Comment by mikeTWC1984** [comment]
OK, so that comes from socket.io (websocket) library. 
 - Do you use all default configs in config.json?
 -  which OS (win 10, server 2016, ..)
 - Do you happen to have VPN or 3rd party software on this machine?

One more thing, try to downgrade to 1.7.0 and see if there are any changes

I can't reproduce it on Win 10. Will try server 16 / 19 tomorrow

**Comment by zslp2** [comment]
Yes, I am using the stock conf\config.json.

Snippet from systeminfo.
```powershell
OS Name:                   Microsoft Windows Server 2022 Standard
OS Version:                10.0.20348 N/A Build 20348
``` 

List of installed software; also installed are Miniforge3 v23.3.1-1 and Notepad++. This is a new system so there isn't much installed beyond corporate basics for a server.
```powershell
(cronicle) PS C:\ProgramData\CronicleEdge> Get-WmiObject -Class Win32_Product | Select-Object -Property Name | Sort

Name
----
Microsoft Visual C++ 2022 X86 Minimum Runtime - 14.36.32532
Microsoft Visual C++ 2022 X86 Additional Runtime - 14.36.32532
Microsoft Monitoring Agent
Configuration Manager Client
Microsoft Visual C++ 2022 X64 Minimum Runtime - 14.36.32532
Microsoft Policy Platform
EMC Avamar for Windows
Microsoft Visual C++ 2022 X64 Additional Runtime - 14.36.32532
Windows Firewall Configuration Provider
UniversalForwarder
HID Global Crescendo Minidriver x64
VMware Tools
``` 

Same error with 1.7.0; different line number but same line of code.
```powershell
TypeError: Cannot read properties of undefined (reading 'listeners')
    at _Server.attach (C:\ProgramData\CronicleEdge\bin\cronicle.js:88136:35)
    at attach (C:\ProgramData\CronicleEdge\bin\cronicle.js:89263:14)
    at _Server.initEngine (C:\ProgramData\CronicleEdge\bin\cronicle.js:92976:43)
    at _Server.attach (C:\ProgramData\CronicleEdge\bin\cronicle.js:92927:14)
    at __construct.startSocketListener (C:\ProgramData\CronicleEdge\bin\cronicle.js:97191:17)
    at __construct.startup (C:\ProgramData\CronicleEdge\bin\cronicle.js:105807:14)
    at C:\ProgramData\CronicleEdge\bin\cronicle.js:10306:18
    at C:\ProgramData\CronicleEdge\bin\cronicle.js:1328:18
    at replenish (C:\ProgramData\CronicleEdge\bin\cronicle.js:438:15)
    at iterateeCallback (C:\ProgramData\CronicleEdge\bin\cronicle.js:422:15)
``` 

The server in question is a corporate device so I am not sure what all configuration and/or policies are applied. Oddly enough, I was able to install v1.7.0 (updated to v1.7.1) on another system back in late August / early September. I am a little nervous about taking that system down for system patches at this point.

**Comment by zslp2** [comment]
Had to recycle the system that I installed some weeks back and everything started back. So I decided to copy that installation over to the current host and cronicle started; I deleted data/ then reinitialized storage. So it seems like I cannot build/bundle a new installation although I can start one that has already been built.

**Comment by zslp2** [comment]
- I bundled cronicle on a server (hostA) on 13-SEP using v1.7.0 then upgraded to v1.7.1 on about 21-SEP. That instance has been running successfully since then. 
- I bundled cronicle on a new server (hostB) about two weeks ago. I have never been able to get that instance of cronicle to startup, which is what caused me to open this issue.
- If I copy the cronicle installation from hostA to hostB, it runs.
- If I bundle a fresh copy of cronicle (v1.7.0, 1.7.1 or 1.7.2) on hostA, it does not run on either hostA or hostB.

So basically I cannot run a freshly bundled copy of cronicle anywhere. I can run the previously bundled cronicle on different servers. I'm happy to provide zip files of installations; just tell me which one(s).

Sorry for the confusion. Hopefully that makes sense.

**Comment by mikeTWC1984** [comment]
OK, so try to bundle latest cornicle on hostA and share it

**Comment by zslp2** [comment]
[cronicle-1.7.2-dist.zip](https://github.com/cronicle-edge/cronicle-edge/files/12910352/cronicle-1.7.2-dist.zip)

Please see attached. Thanks for the help with this issue.

**Comment by zslp2** [comment]
I was able to bundle v1.7.3 and it runs on hostB. I suspect something with a node package but that is a complete guess.

**Comment by mikeTWC1984** [comment]
> I was able to bundle v1.7.3 and it runs on hostB. I suspect something with a node package but that is a complete guess.

Does it work when you copy it over to hostA?

I can reproduce the issue now (using the zip file you provided), will dig into it

**Comment by mikeTWC1984** [comment]
Sounds like you should be able to bundle 1.7.3 (latest) on hostA. That listener handling was added there. 
https://github.com/cronicle-edge/cronicle-edge/commit/962d2e5e164da95acaa916054dddb51948280145

Sounds like somehow you used latest pixl-server-web package (that introduced those listeners) on 1.7.2 (before this handling was added). It's weird becuase I was able to bundle run 1.7.2 with no issue.

**Comment by zslp2** [comment]
I can bundle 1.7.3 on hostA and hostB; everything works as expected. I have no idea how to explain what was happening, especially since you could not reproduce the issue. I have been suspicious of an npm package update since I could not bundle a fresh copy of what I had bundled a few weeks ago. So I think this is best described as a race condition between code release and npm package (specifically pixl-server-web) version; just with the caveat that it was only reproduceable within my network.

I very much appreciate your help in debugging this issue and the enhancements that you add to the already awesome cronicle.

**Comment by mikeTWC1984** [comment]
Actually, somebody just reported the same issue on the main cronicle. The reason is that npm auto upgrades packages to minor versions (if those have ^x.x.x ) if there is no lock file. So upgrading to pixel-server-web 1.3.22 was kinda breaking if you use an older cronicle. I will include that lock file on my repo to prevent this.
